File: basics/02-the-first-triangle/cpp/main.cpp
// Include standard headers.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Include GLEW. Always include it before gl.h and glfw.h, since it's a bit magic.
#include <GL/glew.h>

// Include glfw
#include <GLFW/glfw3.h>

// 3D Mathematics
#include <glm/glm.hpp>
using namespace glm;

// Include our shaders loader.
#include "shaders.cpp"

int main() {
	// Initialize GLFW
	if ( !glfwInit() )
	{
		fprintf( stderr, "Failed to initialize GLFW\n" );
		return -1;
	}

	glfwWindowHint(GLFW_SAMPLES, 4); // 4x antialiasing
	glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3); // We want OpenGL 3.3
	glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
	glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE); // To make MacOS happy
	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE); // We don't want the old OpenGL

	// Open a window and create its OpenGL context
	GLFWwindow* window;
	window = glfwCreateWindow( 1024, 768, "Tutorial 01", NULL, NULL);
	if (window == NULL ) {
		fprintf(stderr, "Failed to open GLFW window.");
		glfwTerminate();
		return -1;
	}
	glfwMakeContextCurrent(window); // Initialize GLEW
	glewExperimental = true; // Needed in core profile
	if (glewInit() != GLEW_OK) {
		fprintf(stderr, "Failed to initialize GLEW");
		return -1;
	}

	// To draw a triangle, create a Vertex Array Object and set it as
	// the current one. This must be done after creating the window and
	// before any other OpenGL calls.
	GLuint VertexArrayID;
	glGenVertexArrays(1, &VertexArrayID);
	glBindVertexArray(VertexArrayID);

	// An array of 3 vectors which represents 3 vertices
	static const GLfloat g_vertex_buffer_data[] = {
		-1.0f, -1.0f, 0.0f,
		 1.0f, -1.0f, 0.0f,
		 0.0f,  1.0f, 0.0f,
	};

	////////////////////////
	// Prepare the triangle.
	////////////////////////
	GLuint vertexbuffer;

	// Generate 1 buffer, put the resulting identifier in vertexbuffer
	glGenBuffers(1, &vertexbuffer);

	// The following commands will talk about our buffer.
	glBindBuffer(GL_ARRAY_BUFFER, vertexbuffer);
	glBufferData(GL_ARRAY_BUFFER, sizeof(g_vertex_buffer_data), g_vertex_buffer_data, GL_STATIC_DRAW);

	// At this point the window would immediately exit, so wait for a keypress.

	// Create and compile our GLSL program from the shaders.
	GLuint programID = LoadShaders("SimpleVertexShader.glsl", "SimpleFragmentShader.glsl");

	// Make the default clear color dark blue.
	glClearColor(0.0f, 0.0f, 0.4f, 0.0f);

	// Ensure we can capture the escape key being pressed below
	glfwSetInputMode(window, GLFW_STICKY_KEYS, GL_TRUE);
	do {
		// Clear the screen.
		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
		glUseProgram(programID);

		// Draw the triangle.
		glEnableVertexAttribArray(0);
		glBindBuffer(GL_ARRAY_BUFFER, vertexbuffer);
		glVertexAttribPointer(
			0,        // attribute 0. No particular reason for 0 but it must match the layout in the shader.
			3,        // size
			GL_FLOAT, // type
			GL_FALSE, // normalized?
			0,        // stride
			(void*)0  // array buffer offset
		);

		// This actually does the drawing.
		// Starting from vertex 0; 3 vertices total -> 1 triangle.
		glDrawArrays(GL_TRIANGLES, 0, 3);
		glDisableVertexAttribArray(0);

		// Swap buffers
		glfwSwapBuffers(window);
		glfwPollEvents();
	} // Check if the ESC key was pressed or the window was closed
	while( glfwGetKey(window, GLFW_KEY_ESCAPE ) != GLFW_PRESS &&
	glfwWindowShouldClose(window) == 0 );
}

File: basics/02-the-first-triangle/go/main.go
package main

import (
	"runtime"
	"local/shaders"
	"github.com/go-gl/gl/v3.3-core/gl"
	"github.com/go-gl/glfw/v3.2/glfw"
)

func init() {
	// This is needed to arrange that main() runs on the main thread.
	// See documentation for functions that are only allowed to be called from
	// the main thread.
	runtime.LockOSThread()
}

func main() {
	err := glfw.Init()
	if err != nil {
		panic(err)
	}
	defer glfw.Terminate()

	glfw.WindowHint(glfw.Samples, 4) // 4x antialiasing
	glfw.WindowHint(glfw.ContextVersionMajor, 3) // We want OpenGL 3.3
	glfw.WindowHint(glfw.ContextVersionMinor, 3)
	glfw.WindowHint(glfw.OpenGLForwardCompatible, glfw.True) // To make MacOS happy
	glfw.WindowHint(glfw.OpenGLProfile, glfw.OpenGLCoreProfile) // We don't want the old OpenGL

	// Open a window and create its OpenGL context.
	window, err := glfw.CreateWindow(1024, 768, "Tutorial 01", nil , nil)
	if err != nil {
		panic(err)
	}

	// This seems to be needed in Go-GL so things like GenVertexArrays
	// don't crash. Put gl.Init *after* creating the window.
	err = gl.Init()
	if err != nil {
		panic(err)
	}

	// Initialize GLEW.
	window.MakeContextCurrent()

	// To draw a triangle, create a Vertex Array Object and set it as
	// the current one. This must be done after creating the window and
	// before any other OpenGL calls.
	var VertexArrayID uint32
	gl.GenVertexArrays(1, &VertexArrayID)
	gl.BindVertexArray(VertexArrayID)

	// An array of 3 vectors which represent 3 vertices for our triangle.
	g_vertex_buffer_data := []float32{
		-1.0, -1.0, 0.0,
		 1.0, -1.0, 0.0,
		 0.0,  1.0, 0.0,
	}

	////////////////////////
	// Prepare the triangle.
	////////////////////////
	var vertexbuffer uint32
	gl.GenBuffers(1, &vertexbuffer)
	gl.BindBuffer(gl.ARRAY_BUFFER, vertexbuffer)
	gl.BufferData(
		gl.ARRAY_BUFFER,
		len(g_vertex_buffer_data)*4,  // 4 bits per pixel
		gl.Ptr(g_vertex_buffer_data),
		gl.STATIC_DRAW,
	)

	// Create and compile our GLSL program from the shaders.
	ProgramID := shaders.LoadShaders("../cpp/SimpleVertexShader.glsl", "../cpp/SimpleFragmentShader.glsl")

	// Make the default clear color dark blue.
	gl.ClearColor(0.0, 0.0, 0.4, 0.0)

	// Ensure we can capture the escape key being pressed below
	window.SetInputMode(glfw.StickyKeysMode, glfw.True)
	for window.GetKey(glfw.KeyEscape) != glfw.Press && !window.ShouldClose() {
		// Clear screen and use GLSL shaders.
		gl.Clear(gl.COLOR_BUFFER_BIT)
		gl.UseProgram(ProgramID)

		// Draw the triangle.
		gl.EnableVertexAttribArray(0)
		gl.BindBuffer(gl.ARRAY_BUFFER, vertexbuffer)
		var f float32 = 0.0
		gl.VertexAttribPointer(
			0,        // attribute 0. No particular reason for 0, but must match the layout in the shader.
			3,        // size
			gl.FLOAT, // type (GL_FLOAT)
			false,    // normalized?
			0,        // stride
			gl.Ptr(&f),
		)
		gl.DrawArrays(gl.TRIANGLES, 0, 3)
		gl.DisableVertexAttribArray(0)

		// Swap buffers.
		window.SwapBuffers()
		glfw.PollEvents()
	}
}
